lib/cloudera-manager/service.rb

```diff
@@ -1,4 +1,64 @@
 module ClouderaManager
   class Service < BaseResource
+    class RestartFailedException < BaseException; end
+
+    belongs_to :cluster
+    has_many :roles
+    primary_key :name
+    collection_path "clusters/:cluster_id/services"
+    after_find { |s| s.cluster_id = s.clusterRef[:clusterName] }
+
+    BULK_RESTART_SAFE_ROLE_TYPES = [
+      "IMPALAD",
+      "DATANODE",
+      "NODEMANAGER"
+    ]
+
+    RESTART_ORDER = {
+      "YARN" => ["JOBHISTORY", "NODEMANAGER", "RESOURCEMANAGER"],
+      "HDFS" => ["SECONDARYNAMENODE", "HTTPFS", "FAILOVERCONTROLLER", "NFSGATEWAY", "DATANODE", "NAMENODE", "JOURNALNODE"],
```

Comment:
I was writing for the general case (figure we may as well open source this sucker) and figured that the secondary namenode when not in HA is a good one to restart first, but on second consideration that doesn't really make much sense, I'll do the other stuff first. It is robust to there being no roles of that type for a given service.